selfie.c

```diff
@@ -7765,6 +7899,421 @@ char* smt_binary(char* opt, char* op1, char* op2) {
+  }
+
+  // restore the original program state
+  pc = original_pc;
+  imm = original_imm;
+  fetch();
+  decode();
+
+  return merge_location;
+}
+
+void add_mergeable_context(uint64_t* context) {
+  uint64_t* entry;
+
+  entry = smalloc(2 * SIZEOFUINT64STAR);
+
+  *(entry + 0) = (uint64_t) mergeable_contexts;
+  *(entry + 1) = (uint64_t) context;
+
+  mergeable_contexts = entry;
+}
+
+uint64_t* get_mergeable_context() {
+  uint64_t* head;
+
+  if (mergeable_contexts == (uint64_t*) 0)
+    return (uint64_t*) 0;
+
+  head = mergeable_contexts;
+  mergeable_contexts = (uint64_t*) *(head + 0);
+
+  return (uint64_t*) *(head + 1);
+}
+
+void add_waiting_context(uint64_t* context) {
+  uint64_t* entry;
+
+  entry = smalloc(2 * SIZEOFUINT64STAR);
+
+  *(entry + 0) = (uint64_t) waiting_contexts;
+  *(entry + 1) = (uint64_t) context;
+
+  waiting_contexts = entry;
+}
+
+uint64_t* get_waiting_context() {
+  uint64_t* head;
+
+  if (waiting_contexts == (uint64_t*) 0)
+    return (uint64_t*) 0;
+
+  head = waiting_contexts;
+  waiting_contexts = (uint64_t*) *(head + 0);
+
+  return (uint64_t*) *(head + 1);
+}
+
+void add_prologue_start_and_corresponding_merge_location(uint64_t prologue_start, uint64_t merge_location) {
+  uint64_t* entry;
+
+  entry = prologues_and_corresponding_merge_locations;
+
+  // do not add duplicates
+  while (entry) {
+    if (*(entry + 1) == prologue_start)
+      return;
+
+    entry = (uint64_t*) *(entry + 0);
+  }
+
+  entry = smalloc(3 * SIZEOFUINT64STAR);
+
+  *(entry + 0) = (uint64_t) prologues_and_corresponding_merge_locations;
+  *(entry + 1) = (uint64_t) prologue_start;
+  *(entry + 2) = (uint64_t) merge_location;
+
+  prologues_and_corresponding_merge_locations = entry;
+}
+
+uint64_t get_merge_location_from_corresponding_prologue_start(uint64_t prologue_start) {
+  uint64_t* entry;
+
+  entry = prologues_and_corresponding_merge_locations;
+
+  while (entry) {
+    if (*(entry + 1) == prologue_start)
+      return (uint64_t) *(entry + 2);
+
+    entry = (uint64_t*) *(entry + 0);
+  }
+
+  return -1;
+}
+
+uint64_t currently_in_this_procedure(uint64_t prologue_start) {
+  return (get_merge_location_from_corresponding_prologue_start(prologue_start) != (uint64_t) -1);
+}
+
+void merge(uint64_t* active_context, uint64_t* mergeable_context, uint64_t location) {
+  // do not merge if merging is disabled
+  if (merge_enabled == 0) {
+    if (current_mergeable_context != (uint64_t*) 0) {
+      add_mergeable_context(current_mergeable_context);
+      current_mergeable_context = (uint64_t*) 0;
+    }
+
+    return;
+  }
+
+  print("; merging two contexts at ");
+  print_code_context_for_instruction(location);
+  println();
+
+  if (prologues_and_corresponding_merge_locations != (uint64_t*) 0)
+    if (get_pc(active_context) == *(prologues_and_corresponding_merge_locations + 2))
+      // we have finished the recursion (i.e. the program has reached a program location which is not part of any recursion)
```

Comment:
I don't see why you can't merge in some recursion. Can you explain?

Reply:
At the moment I have implemented it in such a way that the merging always happens outside of the entire recursion (if there is any). It is of course possible to merge within a recursion, but you have to be very careful, because you can only merge on the same recursion level (because the paths have to be at exactly the same location).  And the problem is that recursions can be nested arbitrarily, so it is difficult to keep track of which paths are on the same recursion level and could be merged. I have been thinking about this for a while and I think it would take quite a lot of code so that every context knows how deep it is in the recursion. Of course it would be more efficient to merge as soon as possible considering the translation time, but for the formula at the end it should not make much of a difference if you merge a bit later (i.e. outside the recursion) since the merging happens anyway. So the question is whether an earlier merge is worth it to make the engine (much?) more complex.

Reply:
Actually I found an issue today regarding the recursion during some tests. I already know how to fix this, I will take care of it and think about it again if we can/should merge earlier in the recursion.

Reply:
Yes, please do. This is an interesting problem! Knowing if you are at a mergeable location appears to involve keeping track of the RA register values on the stack. I believe two locations can only be merged if they share the same prefix of RA register values.

Reply:
We could also just compare the stack pointer, right? At a mergeable location the stack pointer of both contexts must be the same, if they are in different recursion depths, this is not the case.

Reply:
Maybe. We need to be sure that this is enough. The stack pointers may be the same by accident but the way we got there may still be different, not sure.